opm/simulators/linalg/cuistl/CuSparseMatrix.cpp

```diff
@@ -0,0 +1,327 @@
+    }
+
+    const auto nonzeroValues = getNonZeroValues().data();
+
+    auto rowIndices = getRowIndices().data();
+    auto columnIndices = getColumnIndices().data();
+    T alpha = 1.0;
+    T beta = 1.0;
+    OPM_CUSPARSE_SAFE_CALL(detail::cusparseBsrmv(m_cusparseHandle.get(),
+                                                 detail::CUSPARSE_MATRIX_ORDER,
+                                                 CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                                 m_numberOfRows,
+                                                 m_numberOfRows,
+                                                 m_numberOfNonzeroBlocks,
+                                                 &alpha,
+                                                 m_matrixDescription->get(),
+                                                 nonzeroValues,
+                                                 rowIndices,
+                                                 columnIndices,
+                                                 m_blockSize,
+                                                 x.data(),
+                                                 &beta,
+                                                 y.data()));
+}
+
+template <typename T>
+void
+CuSparseMatrix<T>::usmv(T alpha, const CuVector<T>& x, CuVector<T>& y) const
+{
+    assertSameSize(x);
+    assertSameSize(y);
+    if (blockSize() < 2) {
+        OPM_THROW(
+            std::invalid_argument,
+            "CuSparseMatrix<T>::usmv and CuSparseMatrix<T>::mv are only implemented for block sizes greater than 1.");
+    }
+    const auto numberOfRows = N();
+    const auto numberOfNonzeroBlocks = nonzeroes();
+    const auto nonzeroValues = getNonZeroValues().data();
+
+    auto rowIndices = getRowIndices().data();
+    auto columnIndices = getColumnIndices().data();
+
+    T beta = 1.0;
+    OPM_CUSPARSE_SAFE_CALL(detail::cusparseBsrmv(m_cusparseHandle.get(),
+                                                 detail::CUSPARSE_MATRIX_ORDER,
+                                                 CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                                 numberOfRows,
+                                                 numberOfRows,
+                                                 numberOfNonzeroBlocks,
+                                                 &alpha,
+                                                 m_matrixDescription->get(),
+                                                 nonzeroValues,
+                                                 rowIndices,
+                                                 columnIndices,
+                                                 blockSize(),
+                                                 x.data(),
+                                                 &beta,
+                                                 y.data()));
+}
+
+template <class T>
+template <class M>
+void
+CuSparseMatrix<T>::assertSameSize(const M& x) const
```

Comment:
The typename M is potentially misleading here, it should be a V for vector instead perhaps?